File: packages/javascript/lumis/test/languages-conformance.test.ts
/**
 * The JavaScript half of the shared language-name conformance suite.
 *
 * `fixtures/conformance-languages/cases.json` states what loading and the CLI's
 * download must agree on in every runtime. This asserts the JavaScript
 * implementation against it;
 * `crates/lumis-wasm-runtime/tests/languages_conformance.rs` and
 * `packages/elixir/lumis/test/languages_conformance_test.exs` assert theirs
 * against the same file.
 */
import { readFileSync } from "node:fs";
import { describe, expect, it } from "vitest";
import { expandBundles } from "../src/core/language-names.js";

interface Cases {
  spellings: { groups: string[][] };
  passthrough: { names: string[] };
  deduplication: { bundle: string; alsoNamed: string };
  ordering: { input: string[]; firstIs: string; lastIs: string };
  unknownBundles: { names: string[] };
}

const cases = JSON.parse(
  readFileSync(
    new URL("../../../../fixtures/conformance-languages/cases.json", import.meta.url),
    "utf8",
  ),
) as Cases;

describe("language name conformance", () => {
  it("expands every spelling of a bundle the same way", () => {
    expect(cases.spellings.groups.length).toBeGreaterThanOrEqual(5);

    for (const spellings of cases.spellings.groups) {
      const expected = expandBundles([spellings[0]]);
      expect(expected.length, `${spellings[0]} expanded to nothing`).toBeGreaterThan(0);

      for (const spelling of spellings.slice(1)) {
        expect(expandBundles([spelling]), `${spelling} disagrees with ${spellings[0]}`).toEqual(
          expected,
        );
      }
    }
  });

  it("lets a name that is not a bundle survive expansion", () => {
    for (const name of cases.passthrough.names) {
      expect(expandBundles([name])).toEqual([name]);
    }
  });

  it("keeps a bundle member named twice only once", () => {
    const { bundle, alsoNamed } = cases.deduplication;
    const alone = expandBundles([bundle]);
    expect(alone).toContain(alsoNamed);

    const withRepeat = expandBundles([bundle, alsoNamed, alsoNamed]);

    expect(withRepeat.filter((name) => name === alsoNamed)).toHaveLength(1);
    expect(withRepeat).toHaveLength(alone.length);
  });

  it("keeps the order it was given", () => {
    const { input, firstIs, lastIs } = cases.ordering;

    const expanded = expandBundles(input);

    expect(expanded[0]).toBe(firstIs);
    expect(expanded.at(-1)).toBe(lastIs);
  });

  it("rejects an unknown bundle rather than treating it as a language", () => {
    for (const name of cases.unknownBundles.names) {
      expect(() => expandBundles([name]), `${name} should be rejected`).toThrow(
        new RegExp(`Unknown bundle "${name}"`),
      );
    }
  });
});

File: packages/elixir/lumis/test/languages_conformance_test.exs
defmodule Lumis.LanguagesConformanceTest do
  @moduledoc """
  The Elixir half of the shared language-name conformance suite.

  `fixtures/conformance-languages/cases.json` states what loading and the CLI's
  download must agree on in every runtime. This asserts the Elixir implementation
  against it; `crates/lumis-wasm-runtime/tests/languages_conformance.rs` and
  `packages/javascript/lumis/test/languages-conformance.test.ts` assert theirs
  against the same file.
  """
  use ExUnit.Case, async: true

  @cases "../../../../fixtures/conformance-languages/cases.json"
         |> Path.expand(__DIR__)
         |> File.read!()
         |> Jason.decode!()

  defp expand(names), do: Lumis.Languages.expand_bundles(names)

  test "expands every spelling of a bundle the same way" do
    groups = get_in(@cases, ["spellings", "groups"])
    assert length(groups) >= 5, "the catalog has five bundles to cover"

    for [first | rest] <- groups do
      assert {:ok, expected} = expand([first])
      refute expected == [], "#{first} expanded to nothing"

      for spelling <- rest do
        assert {:ok, ^expected} = expand([spelling]), "#{spelling} disagrees with #{first}"
      end
    end
  end

  test "lets a name that is not a bundle survive expansion" do
    for name <- get_in(@cases, ["passthrough", "names"]) do
      assert {:ok, [^name]} = expand([name])
    end
  end

  test "keeps a bundle member named twice only once" do
    bundle = get_in(@cases, ["deduplication", "bundle"])
    member = get_in(@cases, ["deduplication", "alsoNamed"])

    assert {:ok, alone} = expand([bundle])
    assert member in alone

    assert {:ok, with_repeat} = expand([bundle, member, member])

    assert Enum.count(with_repeat, &(&1 == member)) == 1
    assert length(with_repeat) == length(alone)
  end

  test "keeps the order it was given" do
    input = get_in(@cases, ["ordering", "input"])
    first = get_in(@cases, ["ordering", "firstIs"])
    last = get_in(@cases, ["ordering", "lastIs"])

    assert {:ok, expanded} = expand(input)

    assert List.first(expanded) == first
    assert List.last(expanded) == last
  end

  test "rejects an unknown bundle rather than treating it as a language" do
    for name <- get_in(@cases, ["unknownBundles", "names"]) do
      assert {:error, {:unknown_bundle, ^name}} = expand([name]),
             "#{name} should be rejected as an unknown bundle, naming itself"
    end
  end
end
